### PR TITLE
rework: URI creation and URL helper

### LINKED ISSUE #7123: Bug: $this->request->getUri()->getPath() is not compatible with PSR-7
CodeIgniter4 Version: v4.3.1

`$this->request->getUri()->getPath()` returns the route (URI path relative to baseURL).

```
app.baseURL = 'http://localhost:8888/ci431/public/'
```
When navigating to `http://localhost:8888/ci431/public/test?a=b`,

```php
$uri = $this->request->getUri();
echo (string) $uri;   // "http://localhost:8888/ci431/public/test?a=b" → Okay
echo $uri->getPath(); // "test" → NG. It should be "/ci431/public/test" when following PSR-7
```

Ref https://www.php-fig.org/psr/psr-7/

Related #5930

### PR DESCRIPTION
**Pull request by kenjis**
Needs ~~#7252~~, ~~#7256~~, ~~#7297~~

**Description**
Fixes #7123, #7296, #7643

- Fixes
  - `$this->request->getUri()` returns a **normalized** current URI (`SiteURI`)
    - if `$indexPage` is `index.php`, the URI contains `index.php`
  - [BC] `$this->request->getUri()->getPath()` returns full URI path
  - `current_url(true)` returns the same instance with `$this->request->getUri()`
  - `base_url()` and `site_url()` now support protocol-relative links
  - `ControllerTestTrait::withUri()` updates the Request instance, too. Because the Request instance has the URI instance. See https://forum.codeigniter.com/showthread.php?tid=88213
- use `SiteURIFactory`, `SiteURI`
- add `Services::siteurifactory()`
- add `Services::superglobals()`
- URL helper
  - add `SiteURI::baseUrl()` and `SiteURI::siteUrl()`
  - remove `_get_uri()`
  - `site_url()`, `base_url()`, `current_url()` use  `SiteURI` (`Services::request()->getUri()`)

Before:
before
After:
after

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
All checks passed! 

However, to be honest I am a little concerned about whether the existing Controller/HTTP tests will be broken.

**Comment by kenjis** [comment]
Rebased.

**Comment by kenjis** [comment]
Rebased and updated and added docs.

**Comment by kenjis** [comment]
Rebased.

**Comment by kenjis** [comment]
I would like to include this in 4.4.0 if possible, can anyone review this?

File: system/HTTP/SiteURI.php [hunk 1]
```diff
@@ -145,7 +145,7 @@ private function determineBaseURL(
         $uri = new URI($baseURL);
 
         // Update scheme
-        if ($scheme !== null) {
+        if ($scheme !== null && $scheme !== '') {
```
Comment (MGatner) [inline system/HTTP/SiteURI.php:148]: How about:
```suggestion
        if ($scheme) {
```

File: system/HTTP/SiteURI.php [hunk 2]
```diff
@@ -363,4 +363,67 @@ protected function applyParts(array $parts): void
             $this->password = $parts['pass'];
         }
     }
+
+    /**
+     * For baser_url() helper.
```
Comment (MGatner) [inline system/HTTP/SiteURI.php:368]:
```suggestion
     * For base_url() helper.
```

File: system/HTTP/SiteURI.php [hunk 3]
```diff
@@ -363,4 +363,67 @@ protected function applyParts(array $parts): void
...
+
+        $config            = clone config(App::class);
+        $config->indexPage = '';
+
+        $host = $this->getHost();
+
+        $uri = new self($config, $relativePath, $host, $scheme);
+
+        // Support protocol-relative links
+        if ($scheme === '') {
```
Comment (MGatner) [inline system/HTTP/SiteURI.php:385]: Should this handle `null` as well, like above?

File: system/HTTP/SiteURIFactory.php [hunk 4]
```diff
@@ -42,6 +47,7 @@ public function createFromGlobals(): SiteURI
      * Create the SiteURI object from URI string.
      *
      * @internal Used for testing purposes only.
+     * @testTag
```
Comment (MGatner) [inline system/HTTP/SiteURIFactory.php:50]: 🤗

File: system/Test/FeatureTestTrait.php [hunk 5]
```diff
@@ -265,15 +268,24 @@ public function options(string $path, ?array $params = null)
...
 
         // $path may have a query in it
-        $parts                   = explode('?', $path);
-        $_SERVER['QUERY_STRING'] = $parts[1] ?? '';
+        $path  = URI::removeDotSegments($path);
+        $parts = explode('?', $path);
+        $path  = $parts[0];
+        $query = $parts[1] ?? '';
+
+        $_SERVER['QUERY_STRING'] = $query;
```
Comment (MGatner) [inline system/Test/FeatureTestTrait.php:280]: Any reason to handle this via `Superglobals`?

**Review by MGatner — CHANGES_REQUESTED** [review]
I didn't look too closely at the test updates but the overall PR looks very good. I'm so glad to be moving this out of those helpers. A few comments.

File: system/HTTP/SiteURI.php (hunk 1 above)
Comment (samsonasik) [reply system/HTTP/SiteURI.php:148]: @MGatner truthy check should only for bool value, using exact compare null or empty string is better to know the exact usage

**Comment by MGatner** [comment]
Checked in.

File: system/HTTP/SiteURI.php (hunk 3 above)
Comment (kenjis) [reply system/HTTP/SiteURI.php:385]: No. `null` is the default value for `$scheme`, and `base_url()` must return the baseURL (e.g., `http://example.com/`).
But `''` is for protocol-relative links (`//example.com/img/test.jpg`). It must omit the protocol part.

File: system/Test/FeatureTestTrait.php (hunk 5 above)
Comment (kenjis) [reply system/Test/FeatureTestTrait.php]: No. Refactored.

**Review by MGatner — APPROVED** [review]
This is very good! I'm a little nervous how many test changes there are. It looks like the only real assertion changes are leading `/` which hopefully are just internal and won't affect projects. 

Very nice work; I've wanted this framework-wide site URI handling for a long time.

**Comment by MGatner** [comment]
I see a few outstanding lint issues.

**Comment by kenjis** [comment]
Needs #7826

**Comment by kenjis** [comment]
> I'm a little nervous how many test changes there are. It looks like the only real assertion changes are leading / which hopefully are just internal and won't affect projects.

I expect few tests to fail unless you test the URI by setting the state of dependent objects and/or global variables, as CI4 does.

I don't think this is such an unrealistic expectation, since in a normal web app, site URIs are rarely tested.

**Review by samsonasik — APPROVED** [review]

**Comment by kenjis** [comment]
Thank you for reviews!

**Comment by MGatner** [comment]
🥳

**Pull request merged**